### ISSUE TITLE
Missing spec: CSS Borders and Box Decorations 4

### ISSUE THREAD

**Issue by cdoublev** [w3c/browser-specs#1025]
Sections related to borders and box-decorations in CSS Backgrounds 4 have been extracted into a separate specification: CSS Borders and Box Decorations 4 ([commit](https://github.com/w3c/csswg-drafts/commit/cedb83ae7203b0d1e8d91210591d58e08ecdb2f1)).

Related value/type/property definitions are not crawled by Reffy and curated by Webref anymore. 

I guess the spec entry is missing in `./specs.json`, right? Or is there a reason to not include it?

**Comment by tidoust** [comment]
Just haven't had time to look into it yet. This creates a situation where the list will contain two delta specs of the same spec, which is not something that we had envisioned.

At a minimum, at least one of the tests on `index.json` should fail because the `seriesNext` property of `css-backgrounds-3` cannot point at `css-background-4` and this new `css-borders-4` spec at the same time.

**Comment by cdoublev** [comment]
> At a minimum, at least one of the tests on index.json should fail because the seriesNext property of css-backgrounds-3 cannot point at css-background-4 and this new css-borders-4 spec at the same time.

It fails (at [this test](https://github.com/w3c/browser-specs/blob/6ca98d6c874a1318d7e29afd143d0bdb976e6281/test/specs.js#L172)) because the `seriesPrevious` of `css-borders-4` cannot be found.

Would the following changes be enough?

  - add `splitFrom` (or `extractedFrom`) to the schemas of a spec object
  - add an entry in `./specs.json` for CSS Borders 4 with `splitFrom` (or `extractedFrom`) set to `css-backgrounds-3`
  - filter specs with `spec.shortname === bordersSpec.splitFrom` when computing `seriesPrevious`

I cannot test it because there is a recurring fetch error when determining spec filenames while building the spec index.

**Comment by tidoust** [comment]
I think a new `seriesComposition` value is going to be useful to identify the case and avoid breaking guarantees that current consumers may rely on. Some of them may expect delta specs to *always* have a `seriesPrevious` property in particular.

To avoid introducing new properties for something that is hopefully going to remain an exception to the rule, we could perhaps reuse the [`forkOf`](https://github.com/w3c/browser-specs/#forkof) and [`forks`](https://github.com/w3c/browser-specs/#forks) properties that were created to record temporary forks of specifications. Someone may expect that all specs targeted by a `forks` property have a `"fork"` series composition but that should not break much in practice.

In other words, I would propose to:
1. introduce a new `"deltafork"` value for `seriesComposition`;
2. use `forkOf` and `forks` to associate the delta fork entry with the entry it is extracted from;
3. use this for `css-borders-4`;
4. and release a new major version of browser-specs / web-specs packages. New major version because of the new type of entries, but update should not require anything specific if consumer already checks the `seriesComposition` property.

@dontcallmedom, WDYT?

**Comment by dontcallmedom** [comment]
in this model, what would happen to the various series/forks properties when css-borders-4 is no longer a delta?

Intuitively, the right model would be to make seriesNext/seriesPrevious string-or-array, which of course is a more painful change, but may be needed if we want to use the sequence of series to ignore potential duplicated data.

**Comment by tidoust** [comment]
In this model, the entry would look like:

```json
{
  "shortname": "css-borders-4",
  "seriesComposition": "deltafork",
  "series": {
    "shortname": "css-borders"
  },
  "forkOf": "css-backgrounds-3"
}
```

Note that the spec creates its own series right away.

When the spec is no longer a delta, `seriesComposition` changes to `"full"`, `forkOf` disappears. The rest does not change:

```json
{
  "shortname": "css-borders-4",
  "seriesComposition": "full",
  "series": {
    "shortname": "css-borders"
  }
}
```

In the "intuitive" model you describe (I note only `seriesNext` would need to be string-or-array in this model, we could stick to specs linking to only one previous spec through `seriesPrevious`), you'd have:

```json
{
  "shortname": "css-borders-4",
  "seriesComposition": "delta",
  "series": {
    "shortname": "css-backgrounds"
  },
  "seriesPrevious": "css-backgrounds-3"
}
```

The spec would only create its own series when it stops being a delta spec:

```json
{
  "shortname": "css-borders-4",
  "seriesComposition": "full",
  "series": {
    "shortname": "css-borders"
  }
}
```

Question is: when should we consider that the spec create its own series? From a W3C API perspective, the new series will be created right away, or rather as soon as the delta spec is published to /TR. I would prefer to align with that. To achieve that, the "intuitive" model would require doing something clumsy where we both claim that a spec belong to a series (through `seriesPrevious`/`seriesNext` links) but then also claim that it creates its own series at the same time (different `series` property for `css-borders-4` and `css-backgrounds-3`).

Two other non fully satisfactory alternatives that ignore the fact that the spec is a delta spec:
1. Claim that `css-borders-4` is a fork of `css-backgrounds-3`. But then Reffy would get rid of the CSS extract IIRC.
2. Claim that `css-borders-4` is a full spec, no link to `css-backgrounds-3` whatsoever.

The second alternative has the advantage of allowing us to ignore the problem until it really creates a problem (i.e., duplicated data).

**Comment by dontcallmedom** [comment]
isn't there also an issue about dealing with "current" in the series? i.e. if css-backgrounds-3 remains current well after css-borders-4 becomes "full"?

(sorry for not coming up with proposals, I'm still trying to wrap my head around the problem)

**Comment by tidoust** [comment]
Yes, that's also an issue indeed. In Webref, we would have to patch one of the extracts to remove the duplicates, unless specs are considered to be in the same series.

Similarly, `css-backgrounds-4` could become the current spec, while `css-borders-4` remains a delta spec. In Webref, we would have to patch the delta extract with border-related properties from `css-backgrounds-3` since we would no longer have them. That would happen regardless of whether specs are considered to be in the same series or not.

I don't know what to make of this either :)

**Comment by dontcallmedom** [comment]
I guess in my "intuitive" model, I would have used what you qualify as clumsy - a different shortname for the `series` and `seriesPrevious` - what would break if we do that?
```json
{
  "shortname": "css-borders-4",
  "seriesComposition": "delta",
  "series": {
    "shortname": "css-borders"
  },
  "seriesPrevious": "css-backgrounds-3"
}
```

(`css-backgrounds-3` would have to list two `seriesNext`)

**Comment by tidoust** [comment]
What does break mean, and for whom? :)

Regardless of the solution, Reffy will need to be adjusted. Nothing complicated but code follows `seriesNext` links when it adjusts extracts for instance. What will happen for CSS extracts is that Reffy will create a `css/css-borders-4.json` extract simply because we keep delta extracts. It won't create a `css/css-borders.json`. That's the gist of my question: when should we consider that the spec create its own series?

I do not know how other consumers of the list handle series. The guarantee today is that if you take two specs linked through `seriesPrevious`/`seriesNext`, they belong to the same series. A corollary is that the spec identified as current is guaranteed to be in the same series. These guarantees would not longer hold.

Dropping these guarantees may not break existing code. I'm worried that it could break code in subtle ways: it seems easy to miss the fact that there's a separate series hidden in there. That's why I'm advocating being explicit that the spec is a monster, so that code that needs to worry about series knows that it's dealing with a specific case. Said differently, I think it is saner to break things for real than to pretend the spec is business as usual.

If `seriesPrevious` seems better, a middle-ground solution could be to introduce a `"deltabreak"` value for `seriesComposition` and use `seriesPrevious` to link to the previous spec. I'm not even sure that we really need to turn `seriesNext` into a list. That's not super symmetrical but then the new value is there to clarify that the spec is not your usual delta spec.

```json
{
  "shortname": "css-borders-4",
  "seriesComposition": "deltabreak",
  "series": {
    "shortname": "css-borders"
  },
  "seriesPrevious": "css-backgrounds-3"
}
```

`"deltabreak"` isn't exactly a fantastic term. It's meant to convey the fact that, while the spec is part of its own series, it actually is a delta spec of a spec in another series. Value could perhaps be `"deltasplit"` or `"newdelta"`. I had `"deltafork"` in my previous proposal because it was reusing `forkOf`.

**Comment by dontcallmedom** [comment]
how about
```json
{
  "shortname": "css-borders-4",
  "seriesComposition": "delta",
  "series": {
    "shortname": "css-borders"
  }
}
```
(i.e. a `delta` with no `seriesPrevious`)

In a way, this situation isn't very different from CSS3 modules building on top of CSS2, with perhaps the exception that this is starting as a delta; but maybe the "simplest" is not to assume that a delta spec necessarily has a `seriesPrevious`.

> Reffy will create a css/css-borders-4.json extract simply because we keep delta extracts. It won't create a css/css-borders.json. That's the gist of my question: when should we consider that the spec create its own series?

I think they are actually 2 different questions: it seems fine to me that we have css/css-borders-4.json and no css/css-borders.json until css-borders-4 matures in a full spec; the series 'css-borders' will de facto exist, but won't have a matching CSS file (the same as it was a spec without CSS content).

**Comment by dontcallmedom** [comment]
(this leaves open the question of how we would handle a situation where `currentSpecification` and `seriesComposition` don't play well together among these specs; but I think this would have to be be fixed either upstream with the CSS WG, or downstream through patches or using the precedence mechanism)

**Comment by tidoust** [comment]
That's a possibility too and it would break existing code that looks at "delta" specs, sending a clear signal that the logic needs to be updated.

Now this makes me wonder: Why bother recording that the spec is a "delta" spec if we do not record what spec it is a delta of? How could code use that information in a meaningful way?

Said differently, I find it more and more appealing to simply ignore the problem and flag the spec as "full". That's not fully correct of course. But no code change required, and no practical problem that I can think of for now.

**Comment by dontcallmedom** [comment]
the only concern with considering it full when it's only a delta is if the extracted data is messy due to its delta nature - do we know?

**Comment by tidoust** [comment]
> the only concern with considering it full when it's only a delta is if the extracted data is messy due to its delta nature - do we know?

I haven't checked yet, but how would the delta/full nature of the spec change anything for Reffy/Webref? The only difference I can think of is that the CSS extract would be named `css/css-borders-4.json` and not `css/css-borders.json` (all other extracts will be the same ones, well, except if the spec starts defining Web IDL). The CSS extract will ship in `@webref/css` in both cases and we would have to patch messy extracts manually no matter what. Or do you have code that skips over CSS extracts based on the filename?

**Comment by tidoust** [comment]
> Or do you have code that skips over CSS extracts based on the filename?

Actually, we sort of do... At least, the filename convention is explicit in the [guarantees](https://github.com/w3c/webref#css-extracts). While the spec remains a delta, there shouldn't be much duplication issues. When the spec becomes full, we'll have to maintain a patch if `css-backgrounds-3` remains the current specification in the `css-backgrounds` series.

> Should we expect anything bad to happen to dfns with potential duplicates?

I think that we're in usual territory for dfns. Specs that need to reference dfns defined both in css-borders-4 and css-backgrounds-3 would have to decide which spec to reference, and I think that's going to depend on the context (at least, I don't know how to make a decision that would work for all).

**Comment by tidoust** [comment]
Consensus after further live discussion between @dontcallmedom and I: let's record the spec as a full spec for now as that avoids introducing breaking changes and should not create substantial issues, and revisit if the case arises again or if recording the delta aspect of the spec proves necessary.
